src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs

```diff
@@ -1656,22 +1656,51 @@ private void AppendSymbolsWithName(ArrayBuilder<Symbol> results, string name, Bi
 
                     if (!wasError)
                     {
-                        results.Add(singleSymbol);
+                        results.Add(RemapSymbolIfNecessary(singleSymbol));
                     }
                     else
                     {
-                        results.AddRange(lookupResult.Symbols);
+                        foreach (var symbol in lookupResult.Symbols)
+                        {
+                            results.Add(RemapSymbolIfNecessary(symbol));
+                        }
                     }
                 }
                 else
                 {
-                    results.AddRange(lookupResult.Symbols);
+                    foreach (var symbol in lookupResult.Symbols)
+                    {
+                        results.Add(RemapSymbolIfNecessary(symbol));
+                    }
                 }
             }
 
             lookupResult.Free();
         }
 
+        private Symbol RemapSymbolIfNecessary(Symbol symbol)
+        {
+            if (!Compilation.NullableSemanticAnalysisEnabled) return symbol;
+
+            switch (symbol)
+            {
+                case LocalSymbol _:
+                case ParameterSymbol _:
+                case MethodSymbol { MethodKind: MethodKind.LambdaMethod }:
+                case MethodSymbol { MethodKind: MethodKind.LocalFunction }:
+                    return RemapSymbolIfNecessaryCore(symbol, symbol.Locations[0].SourceSpan.Start);
+
+                default:
+                    return symbol;
+            }
+        }
+
+        /// <summary>
+        /// Remaps a local, parameter, localfunction, or lambda symbol, if that symbol or its containing
+        /// symbols were reinferred. This should only be called when nullable semantic analysis is enabled.
+        /// </summary>
+        internal abstract Symbol RemapSymbolIfNecessaryCore(Symbol symbol, int position);
```

Comment:
>int position [](start = 75, length = 12)

Consider removing this parameter and getting location where it is actually needed so that consumer wouldn't have to understand what position is appropriate to pass here.